src/server.c

```diff
@@ -616,14 +617,36 @@ void tryResizeHashTables(int dbid) {
  * The function returns 1 if some rehashing was performed, otherwise 0
  * is returned. */
 int incrementallyRehash(int dbid) {
-    /* Keys dictionary */
-    if (dictIsRehashing(server.db[dbid].dict)) {
-        dictRehashMilliseconds(server.db[dbid].dict,1);
-        return 1; /* already used our millisecond for this loop... */
+    /* Rehash main dictionary. */
+    if (server.cluster_enabled) {
+        listNode *node, *nextNode;
+        monotime timer;
+        elapsedStart(&timer);
+        /* Our goal is to rehash as many slot specific dictionaries as we can before reaching predefined threshold,
+         * while removing those that already finished rehashing from the queue. */
+        while ((node = listFirst(server.db[dbid].rehashing))) {
```

Comment:
using a list holding the dict pointer is very dangerous I think, it's easy to forget clear the pointer in queue when releasing the dict.

the way in `tryResizeHashTables` can be used to rehash safely.

Reply:
I agree it's not ideal and potentially error prone, but I'm afraid that relying on db iterator for rehashing would negatively impact the performance, especially in the case if only few dictionaries are rehashing at the moment. I can try this approach and collect some numbers.